src/bgw/scheduler.c

```diff
@@ -325,19 +330,124 @@ scheduled_bgw_job_start(ScheduledBgwJob *sjob, register_background_worker_callba
 	}
 }
 
-static List *
-initialize_scheduled_jobs_list(MemoryContext mctx)
+static void
+terminate_and_cleanup_job(ScheduledBgwJob *sjob)
 {
-	List	   *jobs = bgw_job_get_all(sizeof(ScheduledBgwJob), mctx);
-	ListCell   *lc;
+	if (sjob->handle != NULL)
+	{
+		TerminateBackgroundWorker(sjob->handle);
+		WaitForBackgroundWorkerShutdown(sjob->handle);
+		pfree(sjob->handle);
+	}
+	sjob->may_need_mark_end = false;
+	worker_state_cleanup(sjob);
+}
 
-	foreach(lc, jobs)
+/*
+ *  Update the given job list with whatever is in the bgw_job table. For overlapping jobs,
+ *  copy over any existing scheduler info from the given jobs list.
+ *  Assume that both lists are ordered by job ID.
+ *  Note that this function call will destroy cur_jobs_list and return a new list.
+ */
+List *
+update_scheduled_jobs_list(List *cur_jobs_list, MemoryContext mctx)
+{
+	List	   *new_jobs = bgw_job_get_all(sizeof(ScheduledBgwJob), mctx);
+	ListCell   *new_ptr = list_head(new_jobs);
+	ListCell   *cur_ptr = list_head(cur_jobs_list);
+
+	while (cur_ptr != NULL)
 	{
-		ScheduledBgwJob *sjob = lfirst(lc);
+		ScheduledBgwJob *new_sjob;
+		ScheduledBgwJob *cur_sjob;
+
+		if (new_ptr == NULL)
+			break;
+
+		new_sjob = lfirst(new_ptr);
+		cur_sjob = lfirst(cur_ptr);
+
+		if (cur_sjob->job.fd.id < new_sjob->job.fd.id)
+		{
+			/*
+			 * We don't need cur_sjob anymore. Make sure to clean up the job
+			 * state. Then keep advancing cur pointer until we catch up.
+			 */
+			while (cur_sjob->job.fd.id < new_sjob->job.fd.id)
+			{
+				/* Terminate and cleanup the job's BGW */
+				terminate_and_cleanup_job(cur_sjob);
+
+				cur_ptr = lnext(cur_ptr);
+				if (cur_ptr == NULL)
+					goto finish_new_list;
+				cur_sjob = lfirst(cur_ptr);
+			}
+		}
+		if (cur_sjob->job.fd.id == new_sjob->job.fd.id)
+		{
+			/*
+			 * Then this job already exists. Copy over any state and advance
+			 * both pointers.
+			 */
+			cur_sjob->job = new_sjob->job;
+			*new_sjob = *cur_sjob;
+
+			cur_ptr = lnext(cur_ptr);
+			new_ptr = lnext(new_ptr);
+		}
+		else if (cur_sjob->job.fd.id > new_sjob->job.fd.id)
+		{
+			/* Advance the new_job list until we catch up to cur_list */
+			new_ptr = lnext(new_ptr);
+		}
+	}
+
+	/* If there's more stuff in cur_list, clean it all up */
+	if (cur_ptr != NULL)
+	{
+		ListCell   *ptr;
 
-		scheduled_bgw_job_transition_state_to(sjob, JOB_STATE_SCHEDULED);
+		for_each_cell(ptr, cur_ptr)
+			terminate_and_cleanup_job(lfirst(ptr));
 	}
-	return jobs;
+
+finish_new_list:
+	if (new_ptr != NULL)
+	{
+		/* Then there are more new jobs. Initialize all of them. */
+		ListCell   *ptr;
+
+		for_each_cell(ptr, new_ptr)
+			scheduled_bgw_job_transition_state_to(lfirst(ptr), JOB_STATE_SCHEDULED);
+	}
+
+	/* Free the old list */
+	list_free(cur_jobs_list);
+	return new_jobs;
+}
+
+void
+populate_scheduled_job_tuple(ListCell *sjob_lc, Datum *values)
```

Comment:
please add a comment that this is only used by test code. Also the parameter should probably be a ScheduledBgwJob *, not a ListCell *

Reply:
Added comment. I didn't make the parameter `ScheduledBgwJob`, because `ScheduledBgwJob` is a static struct defined with the *.c file. I didn't want to expose the struct (hence the helper function).